### PR TITLE
feat(config): Add cryostat origin to APIServer CORS allowed origins

### LINKED ISSUE #320: Add cryostat origin to APIServer CORS allowed origins
In order for the frontend to access the OAuth server logout endpoint, the APIServer needs the cryostat web host added to the spec `additionalCORSAllowedOrigins`. The [docs](https://docs.openshift.com/container-platform/4.8/rest_api/config_apis/apiserver-config-openshift-io-v1.html) mention this "may be needed to access the API and the integrated OAuth server from JavaScript applications". If I don't include this spec, the /logout request fails with a CORS error.
Here's a hardcoded example:
```
apiVersion: config.openshift.io/v1
kind: APIServer
metadata:
  ...
spec:
  additionalCORSAllowedOrigins:
  - cryostat-sample-default\.apps\.jalaw0\.lab\.upshift\.rdu2\.redhat\.com
``` 

_Originally posted by @jan-law in https://github.com/cryostatio/cryostat/pull/786#discussion_r773471630_

### PR DESCRIPTION
**Pull request by jan-law**
Fixes #320

When the cryostat CR is created on OpenShift, the operator should add the spec `AdditionalCORSAllowedOrigins` to the `APIServer` to give cryostat access to the OAuth server's kube:admin logout endpoint.

Some examples (mostly tests) that I found on github:
1) https://github.com/openshift/origin/blob/e945cb88da780e21c021b6c8b430454bcfb881cf/test/extended/crdvalidation/apiserver.go
2) https://github.com/openshift/cluster-kube-apiserver-operator/blob/f47fc346bed6617a90716bd6ed7a90163748df57/test/e2e/user_certs_test.go#L300
3) https://github.com/openshift/cluster-kube-apiserver-operator/blob/39a98d67c3b825b9215454a7817ffadb0577609b/pkg/operator/configobservation/apiserver/observe_cors_test.go#L76
4) https://github.com/PavloVaida/library-go/blob/bb0df7c1ead4f147b82d2abb0623d88cedaaa6ca/pkg/operator/configobserver/apiserver/observe_cors.go

Additional Resources
[APIServer docs [config.openshift.io/v1]](https://docs.openshift.com/container-platform/4.8/rest_api/config_apis/apiserver-config-openshift-io-v1.html)
[How to edit the APIServer manually](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.4/html/security/allowing-javascript-based-access-api-server)
Note: `OpenShiftAPIServer` is not the same as `APIServer`. [OpenShiftAPIServer docs](https://docs.openshift.com/container-platform/4.6/rest_api/operator_apis/openshiftapiserver-operator-openshift-io-v1.html)

### CODE REVIEW

File: internal/controllers/common/resource_definitions/resource_definitions.go [hunk 1]
```diff
@@ -976,6 +981,27 @@ func NewConsoleLink(cr *operatorv1beta1.Cryostat, url string) *consolev1.Console
 	}
 }
 
+func SetAdditionalCorsAllowedOrigins(specs *ServiceSpecs) (*configv1.APIServer, error) {
```
Comment (jan-law) [inline internal/controllers/common/resource_definitions/resource_definitions.go:984]: As of now, I can deploy the operator without any errors, but I don't see an `AdditionalCORSAllowedOrigins` spec when I run `oc describe apiserver cluster`. I also don't see any errors in the `cryostat-operator-controller-manager` logs. 

Would greatly appreciate any tips to debug this, thanks!
```
$ oc describe apiserver cluster
Name:         cluster
Namespace:    
Labels:       <none>
Annotations:  include.release.openshift.io/ibm-cloud-managed: true
              include.release.openshift.io/self-managed-high-availability: true
              include.release.openshift.io/single-node-developer: true
              oauth-apiserver.openshift.io/secure-token-storage: true
              release.openshift.io/create-only: true
API Version:  config.openshift.io/v1
Kind:         APIServer
Metadata:
  Creation Timestamp:  2021-12-17T21:24:58Z
  Generation:          5
  Managed Fields:
    API Version:  config.openshift.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:include.release.openshift.io/ibm-cloud-managed:
          f:include.release.openshift.io/self-managed-high-availability:
          f:include.release.openshift.io/single-node-developer:
          f:oauth-apiserver.openshift.io/secure-token-storage:
          f:release.openshift.io/create-only:
      f:spec:
        .:
        f:audit:
          .:
          f:profile:
    Manager:         cluster-version-operator
    Operation:       Update
    Time:            2021-12-17T21:24:58Z
  Resource Version:  2202263
  UID:               a6c97a3c-d651-48a8-80b0-3d10cfb58ef3
Spec:
  Audit:
    Profile:  Default
```

Comment (jan-law) [inline internal/controllers/common/resource_definitions/resource_definitions.go:984]: As of now, I can deploy the operator without any errors, but I don't see an `AdditionalCORSAllowedOrigins` spec when I run `oc describe apiserver cluster`. I also don't see any errors in the `cryostat-operator-controller-manager` logs. 

Would greatly appreciate any tips to debug this, thanks!
```
$ oc describe apiserver cluster
Name:         cluster
Namespace:    
Labels:       <none>
Annotations:  include.release.openshift.io/ibm-cloud-managed: true
              include.release.openshift.io/self-managed-high-availability: true
              include.release.openshift.io/single-node-developer: true
              oauth-apiserver.openshift.io/secure-token-storage: true
              release.openshift.io/create-only: true
API Version:  config.openshift.io/v1
Kind:         APIServer
Metadata:
  Creation Timestamp:  2021-12-17T21:24:58Z
  Generation:          5
  Managed Fields:
    API Version:  config.openshift.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:include.release.openshift.io/ibm-cloud-managed:
          f:include.release.openshift.io/self-managed-high-availability:
          f:include.release.openshift.io/single-node-developer:
          f:oauth-apiserver.openshift.io/secure-token-storage:
          f:release.openshift.io/create-only:
      f:spec:
        .:
        f:audit:
          .:
          f:profile:
    Manager:         cluster-version-operator
    Operation:       Update
    Time:            2021-12-17T21:24:58Z
  Resource Version:  2202263
  UID:               a6c97a3c-d651-48a8-80b0-3d10cfb58ef3
Spec:
  Audit:
    Profile:  Default
```

Comment (jan-law) [reply internal/controllers/common/resource_definitions/resource_definitions.go:984]: Here are some of the other pieces of code that I've tried using the existing `client.Client` in the reconciler. These snippets all compile, don't output any errors in the cryostat operator logs, and also don't update the APIServer spec:
```
// cryostat_controller.go, around line 356
if r.IsOpenShift {
... 
 apiServerConfig := &configv1.APIServer{
            ObjectMeta: metav1.ObjectMeta{
                Name: "cluster",
            },
            Spec: configv1.APIServerSpec{
                AdditionalCORSAllowedOrigins: []string{serviceSpecs.CoreURL.Hostname()},
            },
        }

        _, err = controllerutil.CreateOrUpdate(ctx, r.Client, apiServerConfig, func() error {
            apiServerConfig.Spec = configv1.APIServerSpec{
                AdditionalCORSAllowedOrigins: []string{serviceSpecs.CoreURL.Hostname()},
            }
            return nil
        })
        if err != nil {
            return reconcile.Result{}, err
        }
...
```
```
// cryostat_controller.go, around line 356
if r.IsOpenShift {
... 
     apiServer := &configv1.APIServer{}
     err = r.Client.Get(context.Background(), types.NamespacedName{Name: "cluster"}, apiServer)
     if err != nil && errors.IsNotFound(err) {
         return reconcile.Result{}, err
     }

     apiServerSpecPatch := resources.NewCorsOriginForApiServer(serviceSpecs)
     err = r.Client.Status().Update(context.Background(), apiServer)
     if err != nil {
         return reconcile.Result{}, err
    }
...

// in resource_definitions.go
func NewCorsOriginForApiServer(specs *ServiceSpecs) client.Patch {

     // Allow access to the OAuth server using CORS headers

     apiServerData := &configv1.APIServer{
         ObjectMeta: metav1.ObjectMeta{
             Name: "cluster",
         },
         Spec: configv1.APIServerSpec{
             AdditionalCORSAllowedOrigins: []string{specs.CoreURL.Hostname()},
        },
     }

     return client.MergeFrom(apiServerData.DeepCopy())
 }
```

File: internal/controllers/common/resource_definitions/resource_definitions.go [hunk 2]
```diff
@@ -976,6 +981,27 @@ func NewConsoleLink(cr *operatorv1beta1.Cryostat, url string) *consolev1.Console
...
+	apiServer, err := configClient.APIServers().Get(context.Background(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	apiServer.Spec.AdditionalCORSAllowedOrigins = append(
+		apiServer.Spec.AdditionalCORSAllowedOrigins,
+		specs.CoreURL.Hostname(),
+	)
+
+	return configClient.APIServers().Update(context.Background(), apiServer, metav1.UpdateOptions{})
```
Comment (ebaron) [inline internal/controllers/common/resource_definitions/resource_definitions.go:1002]: I think you should be able to use the existing client from the Cryostat controller to get and update the APIServer object. You just need to register the config/v1 API with the Manager's Scheme here:
https://github.com/cryostatio/cryostat-operator/blob/5c399ff843136198000a1bf0b99837d0c307a744/internal/main.go#L56-L59

Then you should be able to use the APIServer type just like any other built-in Kubernetes resource.

Comment (ebaron) [reply internal/controllers/common/resource_definitions/resource_definitions.go:1002]: I think it also makes sense to move this functionality to cryostat_controller.go, since you'll have the client object already there.

**Review by ebaron — COMMENTED** [review]
Hi Janelle, I think my comment below should help simplify this change a bit.

File: internal/controllers/common/resource_definitions/resource_definitions.go (hunk 2 above)
Comment (jan-law) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: `reqLogger.Info()` should send output to the logs in `oc logs cryostat-operator-controller-manager-78c899f994-9kt9m`, right?
 
The tests seem to notice an error thrown at `r.Client.Get(context.Background(), types.NamespacedName{Name: "cluster"}, apiServer)`, except when I write `reqLogger.Error(err, "Failed to get APIServer config")`, I don't see any errors printed in the logs. Any ideas?

Comment (ebaron) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: I'm seeing the following error when building and deploying your PR branch:
`E0106 19:09:30.129907       1 reflector.go:127] pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156: Failed to watch *v1.APIServer: failed to list *v1.APIServer: apiservers.config.openshift.io is forbidden: User "system:serviceaccount:cryostat-operator-system:cryostat-operator-service-account" cannot list resource "apiservers" in API group "config.openshift.io" at the cluster scope
`

Try it out with: `make deploy IMG=quay.io/ebaron/cryostat-operator:oauth-logout-01`.

You'll need to give get/list/watch/update permissions to the operator with the kubebuilder marker comments here:
https://github.com/cryostatio/cryostat-operator/blob/91179772f4277810d4d82f4f21f0c5c57acf7c43/internal/controllers/cryostat_controller.go#L104-L121

Then run `make manifests`.

Comment (jan-law) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: > Try it out with: make deploy IMG=quay.io/ebaron/cryostat-operator:oauth-logout-01.

Thanks, I can see the permissions error now.

When I ran the following after adding `// +kubebuilder:rbac:groups=config.openshift.io,resources=apiservers,verbs=list;update
`, I see an unrelated permissions error `User \"system:serviceaccount:default:cryostat-operator-service-account\" cannot delete resource \"clusterrolebindings\" in API group \"rbac.authorization.k8s.io\" at the cluster scope"}`:

```
make generate manifests
BUILDAH_FORMAT=docker podman build -t quay.io/jalaw/cryostat-operator:latest .
podman push quay.io/jalaw/cryostat-operator:latest
DEPLOY_NAMESPACE=$(oc project -q) make deploy CORE_IMG="quay.io/jalaw/cryostat:latest" IMG="quay.io/jalaw/cryostat-operator:latest"
```

The permissions to modify `clusterrolebindings` are unchanged: `// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=clusterrolebindings,verbs=create;get;list;update;watch;delete`.

I think I'm missing something. Are these the right commands to build and deploy the operator?

Comment (ebaron) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: Check the generated config/rbac/role.yaml to see if it added the permissions correctly and didn't remove anything it shouldn't. It looks like the following for me:

```diff
diff --git a/config/rbac/role.yaml b/config/rbac/role.yaml
index 49a7f76..16b1166 100644
--- a/config/rbac/role.yaml
+++ b/config/rbac/role.yaml
@@ -26,6 +26,13 @@ rules:
   - selfsubjectaccessreviews
   verbs:
   - create
+- apiGroups:
+  - config.openshift.io
+  resources:
+  - apiservers
+  verbs:
+  - list
+  - update
 - apiGroups:
   - console.openshift.io
   resources:
```

Testing with this marker added:
`// +kubebuilder:rbac:groups=config.openshift.io,resources=apiservers,verbs=list;update`

I see the following error in the logs:
`E0106 20:43:18.250371       1 reflector.go:127] pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156: Failed to watch *v1.APIServer: unknown (get apiservers.config.openshift.io)`

In addition to `list` and `update`, be sure to add `get` and `watch`.

Comment (jan-law) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: Hmm, `config/rbac/role.yaml` looks okay. I'm still getting the same `cannot delete resource \"clusterrolebindings\"` error.
```
diff --git a/config/rbac/role.yaml b/config/rbac/role.yaml
index 49a7f76..7b35ffe 100644
--- a/config/rbac/role.yaml
+++ b/config/rbac/role.yaml
@@ -26,6 +26,15 @@ rules:
   - selfsubjectaccessreviews
   verbs:
   - create
+- apiGroups:
+  - config.openshift.io
+  resources:
+  - apiservers
+  verbs:
+  - get
+  - list
+  - update
+  - watch
 - apiGroups:
   - console.openshift.io
   resources:
```

Comment (ebaron) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: That's weird, it's working for me now. The only time the operator will try to delete a cluster role binding will be when a Cryostat CR is deleted. Does `make destroy_cryostat_cr` hang for you? If not, does using my build change anything: `quay.io/ebaron/cryostat-operator:oauth-logout-02`?

Comment (jan-law) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: > Does make destroy_cryostat_cr hang for you?

No, I can delete the cryostat CR without problems.
```
kubectl delete -f config/samples/operator_v1beta1_cryostat.yaml
cryostat.operator.cryostat.io "cryostat-sample" deleted
```

> If not, does using my build change anything: quay.io/ebaron/cryostat-operator:oauth-logout-02?

Using your build works well. I'm not sure what could be different between my build and yours?

Comment (ebaron) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: I'm not sure. I modified the Makefile to remove the `test-envtest` requirement from the `oci-build` target and built with `make oci-build IMG=quay.io/ebaron/cryostat-operator:oauth-logout-02`, followed by pushing to Quay with podman. Is this similar to how you're building?

Comment (ebaron) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: What image tag are you using for your development builds?

Comment (jan-law) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: Ah, I think it's my unstable wifi/vpn connection. I've been using the same build steps as you, but this time I deleted all of my pods and tried again. Something must have been hanging from the last time my vpn disconnected, even thought it said the cryostat CR pod was deleted. Thanks!

I still need to add a function to remove the cryostat origin url once the cryostat CR is deleted. When that's working, I'll open the PR.

Comment (ebaron) [reply internal/controllers/common/resource_definitions/resource_definitions.go]: Sounds good!

File: internal/controllers/cryostat_controller_test.go [hunk 3]
```diff
@@ -590,6 +599,12 @@ var _ = Describe("CryostatController", func() {
 						err := t.Client.Get(context.Background(), types.NamespacedName{Name: expectedLink.Name}, link)
 						Expect(kerrors.IsNotFound(err)).To(BeTrue())
 					})
+					It("should remove the application url from APIServer AdditionalCORSAllowedOrigins", func() {
+						apiServer := &configv1.APIServer{}
+						err := t.Client.Get(context.Background(), types.NamespacedName{Name: "cluster"}, apiServer)
+						Expect(kerrors.IsNotFound(err)).To(BeFalse())
```
Comment (ebaron) [inline internal/controllers/cryostat_controller_test.go:605]: I think in this case we should expect err to not be nil.

File: internal/controllers/cryostat_controller_test.go [hunk 4]
```diff
@@ -590,6 +599,12 @@ var _ = Describe("CryostatController", func() {
...
+						Expect(len(apiServer.Spec.AdditionalCORSAllowedOrigins)).To(Equal(int(0)))
```
Comment (ebaron) [inline internal/controllers/cryostat_controller_test.go:606]: You can shorten this with the BeEmpty matcher: https://onsi.github.io/gomega/#beempty

File: go.sum [hunk 5]
```diff
@@ -18,23 +18,19 @@ github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78/go.mod h1:LmzpDX
...
-github.com/Azure/go-autorest/autorest/date v0.2.0 h1:yW+Zlqf26583pE43KhfnhFcdmSWlm5Ew6bxipnr/tbM=
 github.com/Azure/go-autorest/autorest/date v0.2.0/go.mod h1:vcORJHLJEh643/Ioh9+vPmf1Ij9AEBM5FuBIXLmIy0g=
 github.com/Azure/go-autorest/autorest/date v0.3.0 h1:7gUk1U5M/CQbp9WoqinNzJar+8KY+LPI6wiWrP/myHw=
 github.com/Azure/go-autorest/autorest/date v0.3.0/go.mod h1:BI0uouVdmngYNUzGWeSYnokU+TrmwEsOqdt8Y6sso74=
 github.com/Azure/go-autorest/autorest/mocks v0.1.0/go.mod h1:OTyCOPRA2IgIlWxVYxBee2F5Gr4kF2zd2J5cFRaIDN0=
 github.com/Azure/go-autorest/autorest/mocks v0.2.0/go.mod h1:OTyCOPRA2IgIlWxVYxBee2F5Gr4kF2zd2J5cFRaIDN0=
-github.com/Azure/go-autorest/autorest/mocks v0.3.0 h1:qJumjCaCudz+OcqE9/XtEPfvtOjOmKaui4EOpFI6zZc=
 github.com/Azure/go-autorest/autorest/mocks v0.3.0/go.mod h1:a8FDP3DYzQ4RYfVAxAN3SVSiiO77gL2j2ronKKP0syM=
 github.com/Azure/go-autorest/autorest/mocks v0.4.1 h1:K0laFcLE6VLTOwNgSxaGbUcLPuGXlNkbVvq4cW4nIHk=
 github.com/Azure/go-autorest/autorest/mocks v0.4.1/go.mod h1:LTp+uSrOhSkaKrUy935gNZuuIPPVsHlr9DSOxSayd+k=
```
Comment (ebaron) [inline go.sum:36]: These changes seem to be unrelated. Could you revert this file?

File: internal/test/resources.go [hunk 6]
```diff
@@ -1554,3 +1555,14 @@ func NewConsoleLink() *consolev1.ConsoleLink {
...
 	}
 }
+
+func NewApiServer() *configv1.APIServer {
+	return &configv1.APIServer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster",
+		},
+		Spec: configv1.APIServerSpec{
+			AdditionalCORSAllowedOrigins: []string{},
```
Comment (ebaron) [inline internal/test/resources.go:1565]: I think it would be useful to have a test case that checks the behaviour with another entry present in `AdditionalCORSAllowedOrigins`. This would just ensure that we don't interfere with user-specified values here.

File: internal/controllers/cryostat_controller.go [hunk 7]
```diff
@@ -659,6 +676,66 @@ func (r *CryostatReconciler) deleteConsoleLink(ctx context.Context, cr *operator
 	return nil
 }
 
+func (r *CryostatReconciler) addCorsAllowedOriginIfNotPresent(allowedOrigin string, cr *operatorv1beta1.Cryostat) error {
+	reqLogger := r.Log.WithValues("Request.Namespace", cr.Namespace, "Request.Name", cr.Name)
+	apiServer := &configv1.APIServer{}
+	err := r.Client.Get(context.Background(), types.NamespacedName{Name: "cluster"}, apiServer)
```
Comment (ebaron) [inline internal/controllers/cryostat_controller.go:682]: Maybe we should extract the "cluster" string into a constant for these two methods. It might be handy in case this name changes in the future.

**Review by ebaron — COMMENTED** [review]
Hi Janelle, really well done! I have a few minor comments below. Have you noticed any disruption to the cluster with this change? My CRC keeps going down, but I'm not sure if it's related to editing the APIServer config or just being temperamental.

File: go.sum (hunk 5 above)
Comment (jan-law) [reply go.sum]: Whoops, should be fixed now.

File: internal/controllers/cryostat_controller.go (hunk 7 above)
Comment (jan-law) [reply internal/controllers/cryostat_controller.go]: I stored the constant in `resource_definitions`. Let me know if you think of a better place for it.

For future reference, the "cluster" string name is mentioned in the config.openshift.io docs here: https://docs.openshift.com/container-platform/4.6/rest_api/config_apis/apiserver-config-openshift-io-v1.html

**Comment by jan-law** [comment]
> Have you noticed any disruption to the cluster with this change?

I know `2b521e9` was working on Friday and am not sure if `aa85670` was also working then. Today, I'm running into this error when deploying the operator. Looking into it.
```
oc logs -p cryostat-operator-controller-manager-78c899f994-kwtzk 
2022-01-10T22:57:53.892Z	ERROR	controller-runtime.manager	Failed to get API Group-Resources	{"error": "Get \"https://172.30.0.1:443/api?timeout=32s\": dial tcp 172.30.0.1:443: connect: connection refused"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
sigs.k8s.io/controller-runtime/pkg/manager.New
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/manager/manager.go:317
main.main
	/workspace/internal/main.go:92
runtime.main
	/usr/local/go/src/runtime/proc.go:204
2022-01-10T22:57:53.892Z	ERROR	setup	unable to start manager	{"error": "Get \"https://172.30.0.1:443/api?timeout=32s\": dial tcp 172.30.0.1:443: connect: connection refused"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
main.main
	/workspace/internal/main.go:102
runtime.main
	/usr/local/go/src/runtime/proc.go:204
```

File: internal/controllers/cryostat_controller.go [hunk 8]
```diff
@@ -659,6 +676,66 @@ func (r *CryostatReconciler) deleteConsoleLink(ctx context.Context, cr *operator
...
+			return nil
+		}
+	}
+
+	apiServer.Spec.AdditionalCORSAllowedOrigins = append(
+		apiServer.Spec.AdditionalCORSAllowedOrigins,
+		allowedOriginAsRegex,
+	)
+
+	err = r.Client.Update(context.Background(), apiServer)
```
Comment (jan-law) [inline internal/controllers/cryostat_controller.go:701]: I noticed this `Update` call is also adding empty fields for `spec.ClientCA.Name`, `spec.encryption` and `spec.servingCerts`. The docs mention the APIServer will use default behaviour if these fields are "unset", which might be different than having an empty value.

I also tried manually editing the `APIServer` to remove the empty fields. Sometimes I get the below error, and sometimes I can remove the fields. Removing the fields manually has no effect on the pending cryostat pod.
```
jalaw@localhost ~/cryostat-operator (oauth-logout)> oc edit apiserver cluster
Error from server (InternalError): an error on the server ("") has prevented the request from succeeding (get apiservers.config.openshift.io cluster)
```
APIServer spec before running `make create_cryostat_cr`:
```
oc get apiserver cluster -o yaml
...
spec:
  audit:
    profile: Default
```

APIServer spec after running `make create_cryostat_cr`:
```
oc get apiserver cluster -o yaml
...
spec:
  additionalCORSAllowedOrigins:
  - https://cryostat-sample-default\.apps\.jalaw1\.lab\.upshift\.rdu2\.redhat\.com
  audit:
    profile: Default
  clientCA:
    name: ""
  encryption: {}
  servingCerts: {}
```

Comment (jan-law) [reply internal/controllers/cryostat_controller.go:701]: I wonder if the empty spec fields are causing some of these errors below?

while running `make deploy`:
```
error: You must be logged in to the server (Unauthorized)
error: couldn't get available api versions from server: Unauthorized
make: *** [Makefile:175: check_cert_manager] Error 1

$ oc login
Error from server (InternalError): Internal error occurred: unexpected response: 400
$ oc login --loglevel=10
...
server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "Internal error occurred: unexpected response: 400",
  "reason": "InternalError",
  "details": {
    "causes": [
      {
        "message": "unexpected response: 400"
      }
    ]
  },
  "code": 500
}]
F0111 13:05:20.259012  149235 helpers.go:115] Error from server (InternalError): Internal error occurred: unexpected response: 400
...
```
```
oc get pods -w
NAME                                                    READY   STATUS    RESTARTS   AGE
cryostat-operator-controller-manager-78c899f994-qhtdm   1/1     Running   7          68m
cryostat-sample-86468b74d6-xcjj2                        0/3     Pending   0          47s
^C⏎                                                                                                                                                                                                                                                              jalaw@localhost ~/cryostat-operator (oauth-logout) [1]> oc logs -p cryostat-operator-controller-manager-78c899f994-qhtdm
2022-01-11T19:01:56.920Z	ERROR	controller-runtime.manager	Failed to get API Group-Resources	{"error": "Get \"https://172.30.0.1:443/api?timeout=32s\": dial tcp 172.30.0.1:443: connect: connection refused"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
sigs.k8s.io/controller-runtime/pkg/manager.New
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/manager/manager.go:317
main.main
	/workspace/internal/main.go:92
runtime.main
	/usr/local/go/src/runtime/proc.go:204
2022-01-11T19:01:56.920Z	ERROR	setup	unable to start manager	{"error": "Get \"https://172.30.0.1:443/api?timeout=32s\": dial tcp 172.30.0.1:443: connect: connection refused"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
main.main
	/workspace/internal/main.go:102
runtime.main
	/usr/local/go/src/runtime/proc.go:204
```

Comment (ebaron) [reply internal/controllers/cryostat_controller.go:701]: FWIW, I tried this PR on a test cluster other than CRC, and it didn't cause any outages or errors.

File: internal/controllers/cryostat_controller.go (hunk 7 above)
Comment (ebaron) [reply internal/controllers/cryostat_controller.go]: I would just put the constant in `cryostat_controller.go` and keep it private visibility (starts with lower case). It's not needed elsewhere I think?

File: internal/controllers/cryostat_controller.go (hunk 8 above)
Comment (ebaron) [inline internal/controllers/cryostat_controller.go:701]: I just noticed this one minor thing. Could you pass the `context.Context` to this method instead of using `context.Background()` for the update? The context parameter should always be first to make golint happy: https://github.com/golang/go/wiki/CodeReviewComments#contexts.

Comment (jan-law) [reply internal/controllers/cryostat_controller.go]: Thanks, all the errors disappeared after switching to cluster-bot.

File: internal/controllers/cryostat_controller.go (hunk 7 above)
Comment (jan-law) [reply internal/controllers/cryostat_controller.go]: Sounds good

File: internal/test/resources.go (hunk 6 above)
Comment (jan-law) [reply internal/test/resources.go]: I've added an extra entry here and updated the tests to confirm that this extra entry is not affected by the cryostat controller

File: internal/controllers/cryostat_controller_test.go (hunk 3 above)
Comment (jan-law) [reply internal/controllers/cryostat_controller_test.go]: I think the error should be nil here, because we expect the APIServer to remain present after the cryostat CR is deleted, right?

Comment (ebaron) [reply internal/controllers/cryostat_controller_test.go]: Oops, yes I meant it should be nil.

File: internal/controllers/cryostat_controller_test.go (hunk 4 above)
Comment (jan-law) [reply internal/controllers/cryostat_controller_test.go]: After adding an extra entry to the `AdditionalCORSAllowedOrigins` in resources.go, I changed this test case to confirm that the extra entry is still present

**Review by ebaron — APPROVED** [review]
Looks awesome!
